### PR TITLE
PMM-6316 replace native scrollbars with custom

### PR DESCRIPTION
**Pull request by lunaticusgreen**
[PMM-6316](https://jira.percona.com/browse/PMM-6316)

https://github.com/Percona-Lab/pmm-submodules/pull/1046

### CODE REVIEW

File: pmm-app/src/pmm-qan/panel/components/Details/Details.tsx [hunk 1]
```diff
@@ -59,53 +60,58 @@ export const DetailsSection: FC = () => {
...
-          )}
-        >
-          <TabPane tab={<span>{Messages.tabs.details.tab}</span>} key={TabKeys.details}>
-            <Metrics databaseType={databaseType} totals={totals} metrics={metrics} loading={metricsLoading} />
-          </TabPane>
-          {showExamplesTab ? (
-            <TabPane tab={<span>{Messages.tabs.examples.tab}</span>} key={TabKeys.examples}>
-              <Example
-                fingerprint={fingerprint}
+      <RSC style={{ height: '600px' }}>
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-qan/panel/components/Details/Details.tsx:63]: Since the height is fixed here can it be a class?

File: pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.styles.ts [hunk 2]
```diff
@@ -6,16 +6,19 @@ export const getStyles = stylesFactory((theme: GrafanaTheme) => {
...
   const textColor = selectThemeVariant(
     { light: (theme.colors as any).gray85, dark: 'rgba(255, 255, 255, 0.8)' },
-    theme.type
+    theme.type,
   );
 
   return {
     /* This will make the table scrollable when it gets too small */
+    scrollableContent: css`
+      padding-bottom: 21px !important;
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.styles.ts:20]: Maybe make this padding number a variable from the theme like `theme.spacing.md`

File: pmm-app/src/pmm-qan/panel/provider/provider.tools.ts [hunk 3]
```diff
@@ -15,67 +14,68 @@ const setFilters = (query) => FILTERS_NAMES.reduce((acc, filterName) => {
   return acc;
 }, {});
 
+interface GrafanaVariables {
```
Comment (tiagomotasantos) [inline pmm-app/src/pmm-qan/panel/provider/provider.tools.ts:17]: Saw these changes were introduced on the other PR as well, I think it's better to merge the other first and rebase on top of PMM-2.0 so no need to introduce these changes here as well

**Comment by tiagomotasantos** [comment]
[
Screenshot 2020-08-07 at 09 46 26
](url)

If you scroll to the right you see that there is a little gap that was introduced here, seems like you are able to scroll more to the right than you should.

File: pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx [hunk 4]
```diff
@@ -216,9 +221,16 @@ export const Table: FC<TableProps> = ({
           ) : null}
           {rows.length && !loading ? (
             <div {...getTableProps()} className="table">
-              <div {...getTableBodyProps()} className={cx('table-body', styles.tableBody(scroll.y))}>
+              <RSC
```
Comment (atymchuk) [inline pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx:224]: If it's a component, then it should follow the naming convention, i.e. `ReactScrollbar`.

File: pmm-app/src/pmm-qan/panel/provider/provider.tools.ts (hunk 3 above)
Comment (atymchuk) [reply pmm-app/src/pmm-qan/panel/provider/provider.tools.ts:17]: 💯 This way there will be less to review.

File: pmm-app/src/pmm-qan/panel/provider/provider.tsx [hunk 5]
```diff
@@ -1,8 +1,7 @@
-import React, { useCallback, useEffect, useState } from 'react';
+import React, { useEffect, useState } from 'react';
```
Comment (atymchuk) [inline pmm-app/src/pmm-qan/panel/provider/provider.tsx:1]: The changes in this file don't seem to have any relationship to the scrollbars. Where do they belong?

File: pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.styles.ts (hunk 2 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.styles.ts:19]: @tiagomotasantos we don't need to do it because it's a quite specific value that we're using only in one place

File: pmm-app/src/pmm-qan/panel/provider/provider.tools.ts (hunk 3 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-qan/panel/provider/provider.tools.ts:17]: Agree, but we need to wait for grafana7 and merge it only then

File: pmm-app/src/pmm-qan/panel/provider/provider.tsx (hunk 5 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-qan/panel/provider/provider.tsx:1]: It is related to grafana upgrade

File: pmm-app/src/pmm-qan/panel/components/Overview/components/DefaultColumns/DefaultColumns.tsx [hunk 6]
```diff
@@ -6,7 +6,7 @@ import {
   COLUMN_WIDTH,
   FIXED_COLUMN_WIDTH,
   MAIN_METRIC_MIN_WIDTH,
-  ROW_NUMBER_COLUMN_WIDTH,
+  ROW_NUMBER_COLUMN_WIDTH, SCROLLBAR_WIDTH,
```
Comment (nicolalamacchia) [inline pmm-app/src/pmm-qan/panel/components/Overview/components/DefaultColumns/DefaultColumns.tsx:9]: the last one should go on another line :)

File: pmm-app/src/pmm-qan/panel/provider/provider.tsx [hunk 7]
```diff
@@ -117,83 +116,65 @@ const actions = {
...
+    } else {
+      const newState = {
+        ...panelState,
+        from: timeRange.from.utc().format('YYYY-MM-DDTHH:mm:ssZ'),
+        to: timeRange.to.utc().format('YYYY-MM-DDTHH:mm:ssZ'),
+        rawTime: {
+          from: timeRange.raw.from,
+          to: timeRange.raw.to,
+        },
+      };
```
Comment (nicolalamacchia) [inline pmm-app/src/pmm-qan/panel/provider/provider.tsx:167]: This can go outside, since it's common between the two branches

File: pmm-app/src/pmm-inventory/Inventory.tsx [hunk 8]
```diff
@@ -17,6 +17,7 @@ export const InventoryPanel = () => {
     <div id="inventory-wrapper">
       <TabsBar>
         {tabs.map((tab, index) => (
+          // @ts-ignore
```
Comment (nicolalamacchia) [inline pmm-app/src/pmm-inventory/Inventory.tsx:20]: Why is this needed?

File: pmm-app/src/pmm-qan/panel/provider/provider.tsx (hunk 7 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-qan/panel/provider/provider.tsx:167]: These changes is for another task, it will fo afrter rebase so don't need to fix it here

File: pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx [hunk 9]
```diff
@@ -54,10 +55,10 @@ export const Table: FC<TableProps> = ({
       const width = getMainColumnWidth(columns.length);
 
       document.querySelectorAll('.table-body .tr>div:nth-child(2)').forEach((element) => {
-        (element as HTMLElement).style.width = `${width}px`;
+        (element as HTMLElement).style['min-width'] = `${width}px`;
       });
       document.querySelectorAll('.table-body .tr').forEach((element) => {
-        (element as HTMLElement).style.width = `${getAllColumnsWidth(width, columns.length)}px`;
+        (element as HTMLElement).style['min-width'] = `${getAllColumnsWidth(width, columns.length)}px`;
```
[the remark this answers is not in the document]
Comment (lunaticusgreen) [reply pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx:61]: I've tried to use refs but it looks like multiple refs that we need to recreate every time when we update data is not a better option from a readability perspective. If you don't mind I would propose to refactor this part when we will start the implementation of resizable columns.

[the remark this answers is not in the document]
Comment (atymchuk) [reply pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx:61]: Agree for the refs.
For the first part: the call of `getAllColumnsWidth(width, columns.length)` does not depend on the `element`, so no need to make this call every time => cache it.

**Review by atymchuk — CHANGES_REQUESTED** [review]

**Review by atymchuk — APPROVED** [review]

**Review by tiagomotasantos — APPROVED** [review]

File: pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx [hunk 10]
```diff
@@ -216,9 +222,16 @@ export const Table: FC<TableProps> = ({
...
           {rows.length && !loading ? (
             <div {...getTableProps()} className="table">
-              <div {...getTableBodyProps()} className={cx('table-body', styles.tableBody(scroll.y))}>
-                {rows.map(RenderRow)}
+              <div
+                {...getTableBodyProps()}
+                className={cx('table-body')}
+                style={{ height: scroll.y - 70 }}
+              >
+                <Scrollbar forceVisible="x" style={{ maxHeight: scroll.y - 70 }}>
```
Comment (nicolalamacchia) [inline pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx:230]: I think you should consider using a constant here instead of the hardcoded 70

**Review by nicolalamacchia — APPROVED** [review]

File: pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx (hunk 10 above)
Comment (lunaticusgreen) [reply pmm-app/src/pmm-qan/panel/components/Overview/components/QanTable/Table.tsx]: Good point, thanks!

**Review by tiagomotasantos — APPROVED** [review]
